### PR TITLE
New: amd-function-arity rule (fixes #38)

### LINKED ISSUE #38: New rule: amd-function-arity
The rule should make sure that when defining a module in the AMD form of define, the function should have the same length (arity) as the array.
**Example:**
valid:

``` js
define(['a', 'b'], function(a, b) { 
  // module code...
});
```

invalid:

``` js
define(['a', 'b'], function(a) {
  // module code that doesn't use b
});
```

``` js
define(['a', 'b'], function(a, b, c) {
 // c is undefined
});
```

I know in some cases requiring a dependency is used for side effects, but I think in most cases it's just a bug, so the edge cases can disable the rule for the line.

### PR DESCRIPTION
**Pull request by platinumazure**
Hope the test coverage is sufficient. :smile:

Oh, I ran into trouble with the `getDependencyNodes` function in `lib/util.js`-- it was restricted only to string literals. That was a surprise to me. I created a new function `getDependencyStringNodes` and had `lib/rules/no-js-extension` call that, and I repurposed `getDependencyNodes` to do the node search without applying the filter (and implemented `getDependencyStringNodes` in terms of that function), to avoid breaking changes while simultaneously improving maintainability. Hope that's okay.

### CODE REVIEW

File: docs/rules/amd-function-arity.md [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+## Rule Details
+
+This rule aims to ensure that AMD callbacks have the correct number of parameters in order to avoid confusion and maintainability issues.
+
+### Options
+
+This rule takes one option, an object with the key `allowExtraDependencies` and a Boolean value (default is `false`).
+
+```json
+    "requirejs/enforce-define": [2, { "allowExtraDependencies": true }]
```
Comment (cvisco) [inline docs/rules/amd-function-arity.md]: wrong rule name there

Comment (cvisco) [inline docs/rules/amd-function-arity.md]: Also I'm not 100% sold on the naming of this option. It's more about allowing require of dependencies for side effects, no?

Comment (platinumazure) [inline docs/rules/amd-function-arity.md]: Open to suggestions. To me it's immaterial- the main purpose of the extra
dependencies is for side effects, but there could be other reasons. And
yes, I'll fix the wrong rule name, oops.

**Comment by platinumazure** [comment]
@cvisco I fixed the docs typo. If you want me to change the option name, I'm open to it, but I don't know what it should be. :smile:

**Comment by cvisco** [comment]
Nice! Thanks for this, @platinumazure.

**Comment by platinumazure** [comment]
Not sure what happened here. @cvisco?

**Comment by cvisco** [comment]
Gotta be a glitch :man_shrugging:...
